src/librustc/traits/error_reporting.rs

```diff
@@ -817,6 +818,57 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
     }
 
+    fn suggest_remove_reference(&self,
+                                obligation: &PredicateObligation<'tcx>,
+                                err: &mut DiagnosticBuilder<'tcx>,
+                                trait_ref: &ty::Binder<ty::TraitRef<'tcx>>) {
+        let ty::Binder(trait_ref) = trait_ref;
+
+        let span = obligation.cause.span;
+        let mut snippet = match self.tcx.sess.codemap().span_to_snippet(span) {
+            Ok(s) => s,
+            Err(_) => String::from(""),
+        };
+
+        let mut refs_number = 0;
+
+        for c in snippet.chars() {
+            if c == '&' {
+                refs_number += 1;
+            }
```

Comment:
```
} else if !c.is_whitespace() {
    break;
}
```
to avoid iterating over the entire snippet (which can be arbitrarily large) and to account for code like `& & & x`.

Reply:
This can be written as

```
let refs_number = snippet.chars().take_while(|c| c == '&' || !c.is_whitespace()).filter(|c| c == '&').count();
```

Reply:
Shouldn't it be:
```
let refs_number = snippet.chars().filter(|c| !c.is_whitespace()).take_while(|c| c == '&').count();
```

Reply:
I made a mistake, you want either & or whitespace, my code gives you only &. Your code would  indeed work.